src/datascience-ui/data-explorer/mainPanel.tsx

```diff
@@ -305,16 +310,59 @@ export class MainPanel extends React.Component<IMainPanelProps, IMainPanelState>
     }
 
     private normalizeRows(rows: JSONArray): ISlickRow[] {
+        // While processing rows we may encounter Inf, -Inf or NaN.
+        // These rows' column types will initially be 'string' or 'object' so
+        // make sure we update the column types
+        // Set of columns to update based on this batch of rows
+        const columnsToUpdate = new Set<string>();
         // Make sure we have an index field and all rows have an item
-        return rows.map((r: any | undefined) => {
+        const normalizedRows = rows.map((r: any | undefined) => {
             if (!r) {
                 r = {};
             }
             if (!r.hasOwnProperty(this.state.indexColumn)) {
                 r[this.state.indexColumn] = uuid();
             }
+            for (let [key, value] of Object.entries(r)) {
```

Comment:
In practice it should be unusual for inf, nan and -inf to appear in user data. But when it happens it usually indicates something is wrong in the training process, so we need to perform this check for every cell in order to accurately flag their presence.

One heuristic we could use is to only do this check if the first column type is Number, or turn on this check only for Tensors. Not sure if strictly necessary though.